File: android/app/src/main/java/com/nearme/app/ui/premium/PremiumComponents.kt
package com.nearme.app.ui.premium

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.nearme.app.models.*

// MARK: - Premium Badge
@Composable
fun PremiumBadge(
    status: PremiumStatus,
    size: BadgeSize = BadgeSize.MEDIUM,
    modifier: Modifier = Modifier
) {
    val backgroundColor = when (status) {
        PremiumStatus.FREE -> MaterialTheme.colorScheme.secondary
        PremiumStatus.TRIAL -> MaterialTheme.colorScheme.tertiary
        PremiumStatus.PREMIUM -> Color(0xFF6366F1) // Premium purple
    }
    
    Text(
        text = status.displayName.uppercase(),
        style = when (size) {
            BadgeSize.SMALL -> MaterialTheme.typography.labelSmall
            BadgeSize.MEDIUM -> MaterialTheme.typography.labelMedium
            BadgeSize.LARGE -> MaterialTheme.typography.labelLarge
        },
        fontWeight = FontWeight.Bold,
        color = Color.White,
        modifier = modifier
            .background(
                color = backgroundColor,
                shape = RoundedCornerShape(size.cornerRadius)
            )
            .padding(size.padding)
    )
}

enum class BadgeSize(
    val cornerRadius: Int,
    val padding: PaddingValues
) {
    SMALL(4, PaddingValues(horizontal = 6.dp, vertical = 2.dp)),
    MEDIUM(6, PaddingValues(horizontal = 8.dp, vertical = 4.dp)),
    LARGE(8, PaddingValues(horizontal = 12.dp, vertical = 6.dp))
}

// MARK: - Task Limit Progress
@Composable
fun TaskLimitProgress(
    status: TaskLimitStatus,
    modifier: Modifier = Modifier
) {
    Card(
        modifier = modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surfaceVariant
        )
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = "Active Tasks",
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                
                if (status.isPremium) {
                    Row(
                        horizontalArrangement = Arrangement.spacedBy(4.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Icon(
                            imageVector = Icons.Default.AllInclusive,
                            contentDescription = "Unlimited",
                            tint = MaterialTheme.colorScheme.primary,
                            modifier = Modifier.size(16.dp)
                        )
                        Text(
                            text = "Unlimited",
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.primary
                        )
                    }
                } else {
                    Text(
                        text = "${status.currentCount}/${status.maxCount}",
                        style = MaterialTheme.typography.bodyMedium,
                        color = if (status.warningThreshold) {
                            MaterialTheme.colorScheme.error
                        } else {
                            MaterialTheme.colorScheme.onSurface
                        }
                    )
                }
            }
            
            if (!status.isPremium) {
                LinearProgressIndicator(
                    progress = status.progressPercentage,
                    modifier = Modifier.fillMaxWidth(),
                    color = if (status.warningThreshold) {
                        MaterialTheme.colorScheme.error
                    } else {
                        MaterialTheme.colorScheme.primary
                    }
                )
                
                when {
                    status.isAtLimit -> {
                        Row(
                            horizontalArrangement = Arrangement.spacedBy(4.dp),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Icon(
                                imageVector = Icons.Default.Warning,
                                contentDescription = "Warning",
                                tint = MaterialTheme.colorScheme.error,
                                modifier = Modifier.size(16.dp)
                            )
                            Text(
                                text = "Task limit reached",
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.error
                            )
                        }
                    }
                    status.warningThreshold -> {
                        Row(
                            horizontalArrangement = Arrangement.spacedBy(4.dp),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Icon(
                                imageVector = Icons.Default.Info,
                                contentDescription = "Info",
                                tint = MaterialTheme.colorScheme.error,
                                modifier = Modifier.size(16.dp)
                            )
                            Text(
                                text = "${status.remainingTasks} task remaining",
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.error
                            )
                        }
                    }
                }
            }
        }
    }
}

// MARK: - Premium Feature Row
@Composable
fun PremiumFeatureRow(
    feature: PremiumFeature,
    isAvailable: Boolean,
    modifier: Modifier = Modifier
) {
    val icon = when (feature) {
        PremiumFeature.UNLIMITED_TASKS -> Icons.Default.AllInclusive
        PremiumFeature.CUSTOM_NOTIFICATION_SOUNDS -> Icons.Default.VolumeUp
        PremiumFeature.DETAILED_NOTIFICATIONS -> Icons.Default.NotificationsActive
        PremiumFeature.ADVANCED_GEOFENCING -> Icons.Default.LocationOn
        PremiumFeature.PRIORITY_SUPPORT -> Icons.Default.HeadsetMic
        PremiumFeature.EXPORT_DATA -> Icons.Default.FileUpload
    }
    
    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(vertical = 4.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Icon(
            imageVector = icon,
            contentDescription = feature.displayName,
            tint = if (isAvailable) {
                MaterialTheme.colorScheme.primary
            } else {
                MaterialTheme.colorScheme.onSurfaceVariant
            },
            modifier = Modifier.size(24.dp)
        )
        
        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(2.dp)
        ) {
            Text(
                text = feature.displayName,
                style = MaterialTheme.typography.titleMedium,
                color = if (isAvailable) {
                    MaterialTheme.colorScheme.onSurface
                } else {
                    MaterialTheme.colorScheme.onSurfaceVariant
                }
            )
            
            Text(
                text = feature.description,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
        
        Icon(
            imageVector = if (isAvailable) {
                Icons.Default.CheckCircle
            } else {
                Icons.Default.Lock
            },
            contentDescription = if (isAvailable) "Available" else "Locked",
            tint = if (isAvailable) {
                Color(0xFF10B981) // Green
            } else {
                MaterialTheme.colorScheme.onSurfaceVariant
            },
            modifier = Modifier.size(20.dp)
        )
    }
}

// MARK: - Upgrade Prompt Card
@Composable
fun UpgradePromptCard(
    onUpgrade: () -> Unit,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier
) {
    Card(
        modifier = modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.primaryContainer
        ),
        border = BorderStroke(
            width = 1.dp,
            color = MaterialTheme.colorScheme.primary.copy(alpha = 0.3f)
        )
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.Top
            ) {
                Column(
                    modifier = Modifier.weight(1f),
                    verticalArrangement = Arrangement.spacedBy(4.dp)
                ) {
                    Text(
                        text = "Upgrade to Premium",
                        style = MaterialTheme.typography.titleMedium,
                        fontWeight = FontWeight.Bold
                    )
                    
                    Text(
                        text = "Unlock unlimited tasks and premium features",
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
                
                IconButton(
                    onClick = onDismiss,
                    modifier = Modifier.size(24.dp)
                ) {
                    Icon(
                        imageVector = Icons.Default.Close,
                        contentDescription = "Dismiss",
                        tint = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }
            
            Row(
                horizontalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                OutlinedButton(
                    onClick = onUpgrade,
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Learn More")
                }
                
                Button(
                    onClick = onUpgrade,
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Upgrade Now")
                }
            }
        }
    }
}

// MARK: - Task Limit Alert Dialog
@Composable
fun TaskLimitAlertDialog(
    onUpgrade: () -> Unit,
    onDismiss: () -> Unit
) {
    AlertDialog(
        onDismissRequest = onDismiss,
        icon = {
            Icon(
                imageVector = Icons.Default.Warning,
                contentDescription = "Warning",
                tint = MaterialTheme.colorScheme.error
            )
        },
        title = {
            Text(
                text = "Task Limit Reached",
                textAlign = TextAlign.Center
            )
        },
        text = {
            Text(
                text = "Free users are limited to 3 active tasks. Upgrade to Premium for unlimited tasks and more features.",
                textAlign = TextAlign.Center
            )
        },
        confirmButton = {
            Button(onClick = onUpgrade) {
                Text("Upgrade to Premium")
            }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) {
                Text("Maybe Later")
            }
        }
    )
}

// MARK: - Feature Lock Overlay
@Composable
fun FeatureLockOverlay(
    feature: PremiumFeature,
    onUpgrade: () -> Unit,
    modifier: Modifier = Modifier
) {
    Card(
        modifier = modifier,
        elevation = CardDefaults.cardElevation(defaultElevation = 8.dp)
    ) {
        Column(
            modifier = Modifier.padding(24.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            Icon(
                imageVector = Icons.Default.Lock,
                contentDescription = "Locked",
                tint = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.size(48.dp)
            )
            
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Text(
                    text = "Premium Feature",
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold
                )
                
                Text(
                    text = feature.displayName,
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                
                Text(
                    text = feature.description,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    textAlign = TextAlign.Center
                )
            }
            
            Button(
                onClick = onUpgrade,
                modifier = Modifier.fillMaxWidth()
            ) {
                Text("Upgrade to Premium")
            }
        }
    }
}

// MARK: - Previews
@Preview(showBackground = true)
@Composable
fun PremiumComponentsPreview() {
    MaterialTheme {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            PremiumBadge(status = PremiumStatus.PREMIUM)
            
            TaskLimitProgress(
                status = TaskLimitStatus(
                    currentCount = 2,
                    maxCount = 3,
                    isPremium = false
                )
            )
            
            PremiumFeatureRow(
                feature = PremiumFeature.UNLIMITED_TASKS,
                isAvailable = false
            )
            
            UpgradePromptCard(
                onUpgrade = {},
                onDismiss = {}
            )
        }
    }
}

File: ios/NearMe/Components/PremiumComponents.swift
import SwiftUI

// MARK: - Premium Badge
struct PremiumBadge: View {
    let status: PremiumStatus
    let size: BadgeSize
    
    enum BadgeSize {
        case small, medium, large
        
        var fontSize: CGFloat {
            switch self {
            case .small: return 10
            case .medium: return 12
            case .large: return 14
            }
        }
        
        var padding: EdgeInsets {
            switch self {
            case .small: return EdgeInsets(top: 2, leading: 6, bottom: 2, trailing: 6)
            case .medium: return EdgeInsets(top: 4, leading: 8, bottom: 4, trailing: 8)
            case .large: return EdgeInsets(top: 6, leading: 12, bottom: 6, trailing: 12)
            }
        }
    }
    
    var body: some View {
        Text(status.displayName.uppercased())
            .font(.system(size: size.fontSize, weight: .bold, design: .rounded))
            .foregroundColor(.white)
            .padding(size.padding)
            .background(
                RoundedRectangle(cornerRadius: size.fontSize)
                    .fill(Color(status.badgeColor))
            )
    }
}

// MARK: - Task Limit Progress
struct TaskLimitProgress: View {
    let status: TaskLimitStatus
    
    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Text("Active Tasks")
                    .font(.subheadline)
                    .foregroundColor(.secondary)
                
                Spacer()
                
                if status.isPremium {
                    HStack(spacing: 4) {
                        Image(systemName: "infinity")
                            .foregroundColor(.blue)
                        Text("Unlimited")
                            .font(.subheadline)
                            .foregroundColor(.blue)
                    }
                } else {
                    Text("\(status.currentCount)/\(status.maxCount)")
                        .font(.subheadline)
                        .foregroundColor(status.warningThreshold ? .orange : .primary)
                }
            }
            
            if !status.isPremium {
                ProgressView(value: status.progressPercentage)
                    .progressViewStyle(LinearProgressViewStyle(tint: status.warningThreshold ? .orange : .blue))
                
                if status.isAtLimit {
                    HStack {
                        Image(systemName: "exclamationmark.triangle.fill")
                            .foregroundColor(.orange)
                        Text("Task limit reached")
                            .font(.caption)
                            .foregroundColor(.orange)
                    }
                } else if status.warningThreshold {
                    HStack {
                        Image(systemName: "info.circle.fill")
                            .foregroundColor(.orange)
                        Text("\(status.remainingTasks) task remaining")
                            .font(.caption)
                            .foregroundColor(.orange)
                    }
                }
            }
        }
        .padding()
        .background(Color(.systemGray6))
        .cornerRadius(12)
    }
}

// MARK: - Premium Feature Row
struct PremiumFeatureRow: View {
    let feature: PremiumFeature
    let isAvailable: Bool
    
    var body: some View {
        HStack(spacing: 12) {
            Image(systemName: feature.icon)
                .font(.title2)
                .foregroundColor(isAvailable ? .blue : .gray)
                .frame(width: 24)
            
            VStack(alignment: .leading, spacing: 2) {
                Text(feature.displayName)
                    .font(.headline)
                    .foregroundColor(isAvailable ? .primary : .gray)
                
                Text(feature.description)
                    .font(.caption)
                    .foregroundColor(.secondary)
                    .multilineTextAlignment(.leading)
            }
            
            Spacer()
            
            if isAvailable {
                Image(systemName: "checkmark.circle.fill")
                    .foregroundColor(.green)
            } else {
                Image(systemName: "lock.fill")
                    .foregroundColor(.gray)
            }
        }
        .padding(.vertical, 4)
    }
}

// MARK: - Upgrade Prompt Card
struct UpgradePromptCard: View {
    let onUpgrade: () -> Void
    let onDismiss: () -> Void
    
    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            HStack {
                VStack(alignment: .leading, spacing: 4) {
                    Text("Upgrade to Premium")
                        .font(.headline)
                        .foregroundColor(.primary)
                    
                    Text("Unlock unlimited tasks and premium features")
                        .font(.subheadline)
                        .foregroundColor(.secondary)
                }
                
                Spacer()
                
                Button(action: onDismiss) {
                    Image(systemName: "xmark")
                        .foregroundColor(.gray)
                        .font(.caption)
                }
            }
            
            HStack(spacing: 12) {
                Button("Learn More") {
                    onUpgrade()
                }
                .buttonStyle(.bordered)
                
                Button("Upgrade Now") {
                    onUpgrade()
                }
                .buttonStyle(.borderedProminent)
            }
        }
        .padding()
        .background(
            LinearGradient(
                colors: [Color.blue.opacity(0.1), Color.purple.opacity(0.1)],
                startPoint: .topLeading,
                endPoint: .bottomTrailing
            )
        )
        .cornerRadius(12)
        .overlay(
            RoundedRectangle(cornerRadius: 12)
                .stroke(Color.blue.opacity(0.3), lineWidth: 1)
        )
    }
}

// MARK: - Feature Lock Overlay
struct FeatureLockOverlay: View {
    let feature: PremiumFeature
    let onUpgrade: () -> Void
    
    var body: some View {
        VStack(spacing: 16) {
            Image(systemName: "lock.fill")
                .font(.largeTitle)
                .foregroundColor(.gray)
            
            VStack(spacing: 8) {
                Text("Premium Feature")
                    .font(.headline)
                    .foregroundColor(.primary)
                
                Text(feature.displayName)
                    .font(.subheadline)
                    .foregroundColor(.secondary)
                
                Text(feature.description)
                    .font(.caption)
                    .foregroundColor(.secondary)
                    .multilineTextAlignment(.center)
                    .padding(.horizontal)
            }
            
            Button("Upgrade to Premium") {
                onUpgrade()
            }
            .buttonStyle(.borderedProminent)
        }
        .padding(24)
        .background(Color(.systemBackground))
        .cornerRadius(16)
        .shadow(radius: 10)
    }
}

// MARK: - Task Limit Alert
struct TaskLimitAlert: View {
    let onUpgrade: () -> Void
    let onDismiss: () -> Void
    
    var body: some View {
        VStack(spacing: 20) {
            Image(systemName: "exclamationmark.triangle.fill")
                .font(.largeTitle)
                .foregroundColor(.orange)
            
            VStack(spacing: 8) {
                Text("Task Limit Reached")
                    .font(.headline)
                    .foregroundColor(.primary)
                
                Text("Free users are limited to 3 active tasks. Upgrade to Premium for unlimited tasks and more features.")
                    .font(.subheadline)
                    .foregroundColor(.secondary)
                    .multilineTextAlignment(.center)
            }
            
            VStack(spacing: 12) {
                Button("Upgrade to Premium") {
                    onUpgrade()
                }
                .buttonStyle(.borderedProminent)
                .frame(maxWidth: .infinity)
                
                Button("Maybe Later") {
                    onDismiss()
                }
                .buttonStyle(.bordered)
                .frame(maxWidth: .infinity)
            }
        }
        .padding(24)
        .background(Color(.systemBackground))
        .cornerRadius(16)
        .shadow(radius: 10)
    }
}

// MARK: - Premium Features List
struct PremiumFeaturesList: View {
    let features: [PremiumFeature]
    let userService: UserService
    
    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text("Premium Features")
                .font(.headline)
                .foregroundColor(.primary)
            
            VStack(spacing: 12) {
                ForEach(features, id: \.rawValue) { feature in
                    PremiumFeatureRow(
                        feature: feature,
                        isAvailable: userService.hasFeature(feature)
                    )
                }
            }
        }
    }
}

// MARK: - Preview Helpers
#if DEBUG
struct PremiumComponents_Previews: PreviewProvider {
    static var previews: some View {
        VStack(spacing: 20) {
            PremiumBadge(status: .premium, size: .medium)
            
            TaskLimitProgress(status: TaskLimitStatus(
                currentCount: 2,
                maxCount: 3,
                isPremium: false
            ))
            
            PremiumFeatureRow(
                feature: .unlimitedTasks,
                isAvailable: false
            )
            
            UpgradePromptCard(
                onUpgrade: {},
                onDismiss: {}
            )
        }
        .padding()
        .previewLayout(.sizeThatFits)
    }
}
#endif